Adding Porch private authenticated registries functionality documentation (#178)
nephio-project/docs: content/en/docs/porch/using-porch/using-external-private-repositories.md

---
title: "Using external private registries"
type: docs
weight: 4
description: ""
---

To enable the Porch function runner to communicate with authenticated private registries, we must:

> 1. Create a kubernetes secret using a docker `config.json` file.

Reviewer: ```suggestion
1. Create a kubernetes secret using a docker *config.json* file.
```
Please use ** for filenames (https://docs.nephio.org/docs/guides/contributor-guides/documentation/#style-guide)

Author: changed to use * instead of ` as per the style guide. apologies i didn't see it before this is my first documentation PR :)

2. Mount this new secret as a volume on the function runner.
3. Provide this secret mount path to the function runner using the argument `--registry-auth-secret-path`

>  A quick way to generate this secret for your use using your docker `config.json` would be to run the following.

Reviewer: ```suggestion
 A quick way to generate this secret for your use using your docker *config.json* would be to run the following.
```

Author: changed to use * instead of ` as per the style guide

```bash
kubectl create secret generic <SECRET_NAME> --from-file=.dockerconfigjson=/path/to/your/config.json --type=kubernetes.io/dockerconfigjson --dry-run=client -o yaml
```

> This should generate a secret template similar to the one below which you can add to the `2-function-runner.yaml` file present on the Porch deployment found [here](https://github.com/nephio-project/catalog/tree/main/nephio/core/porch)

Reviewer: ```suggestion
This should generate a secret template similar to the one below which you can add to the *2-function-runner.yaml* file present on the Porch deployment found [here](https://github.com/nephio-project/catalog/tree/main/nephio/core/porch)
```

Author: changed to use * instead of ` as per the style guide

```yaml
apiVersion: v1
data:
  .dockerconfigjson: <base64-encoded-data>
kind: Secret
metadata:
  creationTimestamp: null
  name: <SECRET_NAME>
type: kubernetes.io/dockerconfigjson
```

> Next you must mount the secret as a volume on the function runner in the `2-function-runner.yaml` as follows

Reviewer: ```suggestion
Next you must mount the secret as a volume on the function runner in the *2-function-runner.yaml* as follows
```

Author: changed to use * instead of ` as per the style guide

```yaml
    volumeMounts:
      - mountPath: /pod-cache-config
        name: pod-cache-config-volume
      - mountPath: /var/tmp/auth-secret
        name: docker-config
        readOnly: true
volumes:
  - name: pod-cache-config-volume
    configMap:
      name: pod-cache-config
  - name: docker-config
    secret:
      secretName: <SECRET_NAME>
```

You may specify your desired `mountPath:` so long as the function runner can access it.

> Note the chosen `mountPath:` should use its own directory if placed in an existing directory so that it does not overwrite access permissions of the existing directory. `e.g.` if you wish to mount on `/var/tmp` you should use `mountPath: /var/tmp/<SUB_DIRECTORY>` etc.

Reviewer: Please use the Hugo note syntax
{{% alert title="Warning" color="primary" %}}
This is a note.
{{% /alert %}}

Author: changed the notes to use Hugo syntax.

> Note the chosen `mountPath:` should use its own directory if placed in an existing directory so that it does not overwrite access permissions of the existing directory. `e.g.` if you wish to mount on `/var/tmp` you should use `mountPath: /var/tmp/<SUB_DIRECTORY>` etc.

Reviewer: ```suggestion
Note the chosen `mountPath:` should use its own directory if placed in an existing directory so that it does not overwrite access permissions of the existing directory. For example if you wish to mount on `/var/tmp` you should use `mountPath: /var/tmp/<SUB_DIRECTORY>` etc.
```

Author: changed instances of e.g. to For example as per your suggestion.

---

> title: "Using external private registries"

Author: amended as per request.

> title: "Using external private registries"

Author: and changed the file name to match

> title: "Using external private registries"

Reviewer 2: Looks good

> title: "Using external private registries"

Reviewer 2: Looks good - OK to resolve

type: docs
weight: 4
description: ""
---

> To enable the Porch function runner to communicate with authenticated private registries, we must:

Author: amended as per request.

> To enable the Porch function runner to communicate with authenticated private registries, we must:

Author: amended as per request.

> To enable the Porch function runner to communicate with authenticated private registries, we must:

Reviewer 2: Looks good - OK to resolve

> To enable the Porch function runner to communicate with authenticated private registries, we must:

Reviewer 2: (another comment on updated version - OK to resolve this one)

> 1. Create a kubernetes secret using a docker *config.json* file.

Author: amended as per request.

> 1. Create a kubernetes secret using a docker *config.json* file.

Reviewer 2: Looks good 👍

2. Mount this new secret as a volume on the function runner.

> 3. Provide this secret mount path to the function runner using the argument `--registry-auth-secret-path`

Author: amended as per request.

> 3. Provide this secret mount path to the function runner using the argument `--registry-auth-secret-path`

Reviewer 2: Looks good - OK to resolve

>  A quick way to generate this secret for your use using your docker *config.json* would be to run the following.

Author: amended as per request.

>  A quick way to generate this secret for your use using your docker *config.json* would be to run the following.

Reviewer 2: Looks good - OK to resolve

```bash
kubectl create secret generic <SECRET_NAME> --from-file=.dockerconfigjson=/path/to/your/config.json --type=kubernetes.io/dockerconfigjson --dry-run=client -o yaml -n porch-system
```

> {{% alert title="Warning" color="primary" %}}

Reviewer: ```suggestion
{{% alert title="Note" color="primary" %}}
```

Aagh sorry, it is a Note, not a Warning.

Author: yes i had a look at the preview there and was wondering why it said warning. ill go make that change now

Author: fixed now

> The secret should be in the same namespace as the function runner deployment which by default is the *porch-system* namespace.

Author: amended as per request.

> The secret should be in the same namespace as the function runner deployment which by default is the *porch-system* namespace.

Author: amended as per request.

> The secret should be in the same namespace as the function runner deployment which by default is the *porch-system* namespace.

Reviewer 2: Looks good - OK to resolve

> The secret should be in the same namespace as the function runner deployment which by default is the *porch-system* namespace.

Reviewer 2: Looks good - OK to resolve

{{% /alert %}}

> This should generate a secret template similar to the one below which you can add to the *2-function-runner.yaml* file present on the Porch deployment found [here](https://github.com/nephio-project/catalog/tree/main/nephio/core/porch)

Reviewer 2: Looks good 👍

```yaml
apiVersion: v1
data:
  .dockerconfigjson: <base64-encoded-data>
kind: Secret
metadata:
  creationTimestamp: null
  name: <SECRET_NAME>
  namespace: porch-system
type: kubernetes.io/dockerconfigjson
```

> Next you must mount the secret as a volume on the function runner in the *2-function-runner.yaml* as follows

Reviewer 2: Looks good 👍

```yaml
    volumeMounts:
      - mountPath: /pod-cache-config
        name: pod-cache-config-volume
      - mountPath: /var/tmp/auth-secret
        name: docker-config
        readOnly: true
volumes:
  - name: pod-cache-config-volume
    configMap:
      name: pod-cache-config
  - name: docker-config
    secret:
      secretName: <SECRET_NAME>
```

You may specify your desired `mountPath:` so long as the function runner can access it.

> {{% alert title="Warning" color="primary" %}}

Reviewer: ```suggestion
{{% alert title="Note" color="primary" %}}
```

Author: fixed also

> The chosen `mountPath:` should use its own directory if placed in an existing directory so that it does not overwrite access permissions of the existing directory. For example if you wish to mount on `/var/tmp` you should use `mountPath: /var/tmp/<SUB_DIRECTORY>` etc.

Author: amended as per request.

> The chosen `mountPath:` should use its own directory if placed in an existing directory so that it does not overwrite access permissions of the existing directory. For example if you wish to mount on `/var/tmp` you should use `mountPath: /var/tmp/<SUB_DIRECTORY>` etc.

Reviewer 2: Looks good 👍

{{% /alert %}}

> Lastly you must add the `--registry-auth-secret-path` to the function runner arguments, giving the path of the secret file mount.

Reviewer 2: Consider "`...to the arguments of the function-runner Deployment object in the *2-function-runner.yaml* file:`, giving the path of the secret file mount:"

Reviewer 2: Looks good - OK to resolve

```yaml
command:
  - /server
  - --config=/config.yaml
  - --registry-auth-secret-path=/var/tmp/auth-secret/.dockerconfigjson
  - --functions=/functions
  - --pod-namespace=porch-fn-system
```

> With this last step, if your Porch package uses a custom kpt function image stored in a private registry (For example `- image: ghcr.io/private-registry/set-namespace:customv2`), the function runner will now use the secret info as an `imagePullSecret` for the function pods as documented [here](https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/).

Author: amended as per request.

> With this last step, if your Porch package uses a custom kpt function image stored in a private registry (For example `- image: ghcr.io/private-registry/set-namespace:customv2`), the function runner will now use the secret info as an `imagePullSecret` for the function pods as documented [here](https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/).

Reviewer 2: Looks good - OK to resolve
